File: code/scenes/src/main/java/org/fbme/scenes/controllers/diagram/ConnectionController.java
package org.fbme.scenes.controllers.diagram;

import jetbrains.mps.nodeEditor.cells.EditorCell;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.function.BiFunction;
import java.util.function.Function;

public interface ConnectionController<CursorT, PathT> {

    Function<Point, PathT> getEdgeTransformation(PathT path, int x, int y);

    Function<Point, PathT> getSourceTransformation(PathT path);

    Function<Point, PathT> getTargetTransformation(PathT path);

    BiFunction<Point, Point, PathT> getEndpointsTransformation(PathT path);

    boolean isSourceTransformableAt(PathT path, int x, int y);

    boolean isTargetTransformableAt(PathT path, int x, int y);

    boolean isSelectableAt(PathT path, int x, int y);

    CursorT getCursorAt(PathT path, int x, int y);

    Rectangle getBounds(PathT path);

    void paintConneciton(PathT path, CursorT cursor, boolean selected, Graphics graphics);

    void paintTrace(PathT path, Graphics graphics);

    @Nullable
    default EditorCell getConnectionCell() {
        return null;
    }

    default void updateCellWithPath(PathT path) {
    }

    default void updateCellSelection(boolean selected) {
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/diagram/ConnectionController.kt
package org.fbme.scenes.controllers.diagram

import jetbrains.mps.nodeEditor.cells.EditorCell
import java.awt.Graphics
import java.awt.Point
import java.awt.Rectangle
import java.util.function.BiFunction
import java.util.function.Function

interface ConnectionController<CursorT, PathT> {
    fun getEdgeTransformation(path: PathT, x: Int, y: Int): Function<Point, PathT>?
    fun getSourceTransformation(path: PathT): Function<Point, PathT>?
    fun getTargetTransformation(path: PathT): Function<Point, PathT>?
    fun getEndpointsTransformation(path: PathT): BiFunction<Point, Point, PathT>?
    fun isSourceTransformableAt(path: PathT, x: Int, y: Int): Boolean
    fun isTargetTransformableAt(path: PathT, x: Int, y: Int): Boolean
    fun isSelectableAt(path: PathT, x: Int, y: Int): Boolean
    fun getCursorAt(path: PathT, x: Int, y: Int): CursorT?
    fun getBounds(path: PathT): Rectangle
    fun paintConnection(path: PathT, cursor: CursorT?, selected: Boolean, graphics: Graphics)
    fun paintTrace(path: PathT, graphics: Graphics)
    val connectionCell: EditorCell?
        get() = null

    fun updateCellWithPath(path: PathT) {}
    fun updateCellSelection(selected: Boolean) {}
}
